rafthttp/snapshot.go

```diff
@@ -0,0 +1,230 @@
+// before the kept snapshot is retrieved.
+// Before a snapshot is retrieved, its raft snapshot message needs to
+// be sent out, and snapHandler waits remote member requesting
+// snapshot data. The process needs to take two maximal RTT, which is 10s.
+// So we set 15s as the timeout.
+const snapshotAutoReleaseTimeout = 15 * time.Second
+
+func newSnapshotKeeper(r Raft) *snapshotKeeper {
+	return &snapshotKeeper{r: r}
+}
+
+func newSnapshotProcessor(tr http.RoundTripper, r Raft, snapSaver SnapshotSaver) *snapshotProcessor {
+	return &snapshotProcessor{
+		tr:        tr,
+		r:         r,
+		snapSaver: snapSaver,
+	}
+}
+
+// snapshotKeeper keeps a snapshot for transporter. transporter
+// needs to attach the member ID that the snapshot is sent to
+// as soon as it knows.
+// The kept snapshot can either be released by snapHandler, or
+// be released automatically when it timeout. If snapshot is released
+// automatically, snapshotKeeper also closes the snapshot and
+// reports snapshot failure to raft.
+// snapshotKeeper keeps at most one snapshot at a time.
+type snapshotKeeper struct {
+	r Raft
+
+	// protect fields below
+	mu sync.Mutex
+	// index is the index of snapshot data for snapshot match
+	// index is set to 0 if there is no kept snapshot
+	index uint64
+	// w is the WriterTo that can write the snapshot data
+	w io.WriterTo
+	// remote is the member ID that the kept snapshot will be sent to
+	// remote is set to raft.None if the kept snapshot has not attached remote
+	remote types.ID
+	// when the closeTimer expires, it will call the function attached
+	// via AfterFunc to close the snapshot
+	// closeTimer is set to nil if the kept snapshot has no auto release
+	closeTimer *time.Timer
+}
+
+// keep keeps the given snapshot at the given index.
+// keep keeps at most one snapshot at a time, or it panics.
+func (s *snapshotKeeper) keep(index uint64, w io.WriterTo) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.index != 0 {
+		plog.Panicf("unexpected keep snapshot when some snapshot is kept")
+	}
+	s.index = index
+	s.w = w
+	s.remote = types.ID(raft.None)
+	s.closeTimer = nil
+}
+
+// attachRemote attaches the given remote member ID to the kept snapshot
+// with the given index.
+// If there is no available snapshot, it panics.
+func (s *snapshotKeeper) attachRemote(index uint64, remote types.ID) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.index != index {
+		plog.Panicf("unexpected attach remote to snapshot at index %d", index)
+	}
+	s.remote = remote
+}
+
+// autoRelease releases the kept snapshot with the given index automatically
+// after the given timeout passes.
+// If the snapshot is released automatically, snapshotKeeper also closes
+// the snapshot, and reports snapshot failure to raft.
+// If there is no available snapshot that has attached remote, it panics.
+func (s *snapshotKeeper) autoRelease(index uint64, timeout time.Duration) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.index != index {
```

Comment:
to detect the potential bug early. If this happens, it indicates that snapshot data cannot match the index, which could let follower recover to mismatch storage.